### PR TITLE
[ty] Make membership and equality narrowing consistent

### LINKED ISSUE #4039: Membership and equality narrowing are inconsistent
### Summary

Here's our current behaviour:

```py
class Foo: ...

def f(x: Foo | None, y: list[Foo], z: Foo):
    if x == z:
        reveal_type(x)  # Foo | None, reflecting the possibility that `z`
                        # could be an instance of a `Foo` subclass that compares equal to `None`
    
    if x in y:
        reveal_type(x)  # Foo -- but `list.__contains__` uses equality semantics internally,
                        # so why should this be different?
```

Here's an example of the kind of thing where this can get us into trouble:

```pycon
>>> class Foo: ...
>>> class Bar(Foo):
...     def __eq__(self, other):
...         return True
...
>>> x: list[Foo] = [Bar()]
>>> None in x
True
```

The bug is caused by the use of `is_single_valued` here: https://github.com/astral-sh/ruff/blob/8bbdb8e392e13481db33abc731546eec5a827d36/crates/ty_python_semantic/src/types/narrow.rs#L3034-L3055 (`None` is a single-valued type). I don't think there's actually any reason why we should be treating single-valued types differently to other types here.

### Version

_No response_

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Prior to this change, membership narrowing removed single-valued union members when the container's element type was broad, even though an element subclass could define custom equality. That made `in` and `==` inconsistent and could unsoundly remove `None`:

```py
class Foo: ...

def f(x: Foo | None, y: Foo, values: list[Foo]):
    if x == y:
        reveal_type(x)  # Foo | None
    if x in values:
        reveal_type(x)  # Previously: Foo
```

This removes the membership-only single-valued fallback and reuses the shared equality evaluator.

Closes https://github.com/astral-sh/ty/issues/4039.

### CODE REVIEW

File: crates/ty_python_semantic/resources/mdtest/narrow/conditionals/in.md [hunk 1]
```diff
@@ -1061,21 +1074,22 @@ def custom_containment_component_prevents_narrowing(
...
 ```
 
 ## `TypedDict` key membership
 
-Membership in a `TypedDict` checks its string keys, so the tested value can be narrowed to a
-possible key. We do not apply key-based narrowing to arbitrary values, because `in` may test
-substrings or elements instead:
+Membership in a `TypedDict` checks its string keys. Equality against a broad `str` key type cannot
+rule out an integer literal. We do not apply key-based narrowing to arbitrary values, because `in`
+may test substrings or elements instead:
```
Comment (AlexWaygood) [inline crates/ty_python_semantic/resources/mdtest/narrow/conditionals/in.md]: one interesting property of a `TypedDict` (as opposed to a `dict[str, object]`) is that we do actually know that all the keys are instances of _exactly_ `str` -- it's not valid for a `TypedDict` to have `str`-subclass keys. An open `TypedDict` can have arbitrary additional keys, so I don't think you can safely say that the keys of an open `TypedDict` are of type `LiteralString` (which is a narrower type than "exactly `str`, not a subclass"). For a `closed=True` `TypedDict`, however, you know that the type of the keys will be a union of all the enumerated `TypedDict` field names in the class definition -- e.g. `Literal["foo", "bar"]` for a `closed=True` `TypedDict` with fields `foo` and `bar`.

File: crates/ty_python_semantic/src/types/narrow.rs [hunk 2]
```diff
@@ -3020,39 +3015,7 @@ impl<'db> NarrowingConstraintsBuilder<'db, '_> {
...
-        let has_single_valued_component = match lhs_ty {
-            Type::Union(union) => union
-                .elements(self.db)
-                .iter()
-                .any(|element| element.is_single_valued(self.db)),
-            _ => lhs_ty.is_single_valued(self.db),
-        };
-        if !has_single_valued_component {
-            return evaluate_type_equality(self.db, lhs_ty, rhs_ty, true, soundness_policy);
-        }
```
Comment (AlexWaygood) [inline crates/ty_python_semantic/src/types/narrow.rs:3032]: I realised in https://github.com/astral-sh/ruff/pull/26981 that you can remove this section without any tests failing, but it is actually load-bearing for some ecosystem inference. A snippet we could add to the test suite to add some coverage for this edge case would be:

```py
class C: ...


def f(origin: C | int):
    if origin in ("x",):
        reveal_type(origin)
```

**Comment by AlexWaygood** [comment]
> while preserving constrained-TypeVar correlation. 

I see the reference to this term in the source code, but I'm unfamiliar with what it means. 

Why is it necessary to do something _different_ for `in` narrowing to "preserve constrained-TypeVar correlation"? Do we not need "constrained-TypeVar correlation" for `==` narrowing?

**Comment by charliermarsh** [comment]
I haven't even looked at this PR myself yet!

**Comment by charliermarsh** [comment]
For reference, "preserve constrained-TypeVar correlation" generally refers to something like this:
```python
def combine[T: (int, str)](x: T, y: T) -> T:
    return x + y
```

In which `(int, int) -> int` and `(str, str) -> str` are valid, but (e.g.) `(int, str) -> int` is not. That is, cases in which the same TypeVar is used multiple times, and so you need to avoid treating them as independent unions to be solved.

**Comment by charliermarsh** [comment]
[PR_26982_ECOSYSTEM_SUMMARY.md](https://github.com/user-attachments/files/30170166/PR_26982_ECOSYSTEM_SUMMARY.md)

**Comment by charliermarsh** [comment]
A lot of new diagnostics, but the report suggests these are expected. An example from CPython:
```python
if self.subclass_of in type_checks:
    typecheck, typename = type_checks[self.subclass_of]
```

Where `self.subclass_of` is `str | None` and `type_checks` is `dict[str, tuple[str, str]]`. Previously, we assumed that the narrowed type is `str`, which was inconsistent with our handling of equality.

**Comment by charliermarsh** [comment]
As discussed in the linked issues, we might want to allow this unsoundness, but we should be consistent about it between `==` and `in` (which this PR addresses).

File: crates/ty_python_semantic/resources/mdtest/narrow/conditionals/in.md [hunk 3]
```diff
@@ -1061,31 +1080,35 @@ def custom_containment_component_prevents_narrowing(
...
+    other: str
 
-def typed_dict_container(value: Literal["present", 1], values: Values) -> None:
+def typed_dict_container(value: Literal["present", "other", "missing", 1], values: Values) -> None:
     if value in values:
-        reveal_type(value)  # revealed: Literal["present"]
+        reveal_type(value)  # revealed: Literal["other", "present"]
 
 def f(x: Literal["abc", "def"]):
     if "a" in x:
```
Comment (AlexWaygood) [inline crates/ty_python_semantic/resources/mdtest/narrow/conditionals/in.md]: I suppose this is a pre-existing issue, but I don't understand why this `f()` function here is in a section titled "TypedDict key membership" -- it doesn't seem to have anything to do with TypedDicts

File: crates/ty_python_semantic/resources/mdtest/narrow/conditionals/in.md [hunk 4]
```diff
@@ -1061,31 +1080,35 @@ def custom_containment_component_prevents_narrowing(
...
 
 ```py
-from typing import Literal, TypedDict
+from typing import Literal
+from typing_extensions import TypedDict
 
-class Values(TypedDict):
+class Values(TypedDict, closed=True):
     present: int
+    other: str
```
Comment (AlexWaygood) [inline crates/ty_python_semantic/resources/mdtest/narrow/conditionals/in.md]: it's good that we have some coverage of `closed=True` TypedDicts here, but shouldn't we also retain some coverage of open `TypedDict`s? (possibly with a TODO comment saying that it would be safe to narrow `int` values away using containment in an open `TypedDict` if we added a concept of "exact `str`, not a subclass of `str`" to our model)

File: crates/ty_python_semantic/src/types/narrow.rs [hunk 5]
```diff
@@ -3020,39 +3015,7 @@ impl<'db> NarrowingConstraintsBuilder<'db, '_> {
             return Some(rhs_ty);
         }
```
Comment (AlexWaygood) [inline crates/ty_python_semantic/src/types/narrow.rs:3016]: This can be deferred to a followup, but there's still an inconsistency between `==` narrowing and `in` narrowing on this branch due to the special case that's being carved out for "correlated TypeVars" in the block of code above this line. In this case, though, I think we want to make the special case more _general_ (move it to `equality.rs`) so that equality-based narrowing also benefits from it:

```py
from enum import Enum
from typing import Literal, TypeVar

class E(Enum):
    A = 1
    B = 2

T = TypeVar("T", Literal[E.A], Literal[E.B])

def correlated_typevar_membership(x: E, y: T) -> T:
    if x in (y,):
        reveal_type(x)  # revealed: T@correlated_typevar
        return x
    return y

def correlated_typevar_eq(x: E, y: T) -> T:
    if x == y:
        reveal_type(x)  # revealed: E
        return x  # false-positive error
    return y
```

**Review by AlexWaygood — APPROVED** [review]

File: crates/ty_python_semantic/src/types/narrow.rs (hunk 5 above)
Comment (charliermarsh) [reply crates/ty_python_semantic/src/types/narrow.rs:3016]: I'll look into it now, probably before merging this.

Comment (AlexWaygood) [reply crates/ty_python_semantic/src/types/narrow.rs:3016]: FWIW if we do generalise the special case by moving it to `equality.rs` somehow, this function will become very trivial:

```diff
diff --git a/crates/ty_python_semantic/src/types/narrow.rs b/crates/ty_python_semantic/src/types/narrow.rs
index 16cd913e3f..c30ced0456 100644
--- a/crates/ty_python_semantic/src/types/narrow.rs
+++ b/crates/ty_python_semantic/src/types/narrow.rs
@@ -2997,25 +2997,13 @@ impl<'db> NarrowingConstraintsBuilder<'db, '_> {
         lhs_ty: Type<'db>,
         rhs_ty: Type<'db>,
     ) -> Option<Type<'db>> {
-        let lhs_ty = lhs_ty.resolve_type_alias(self.db);
-        let rhs_ty = rhs_ty.resolve_type_alias(self.db);
-        let soundness_policy = self.comparison_soundness_policy();
-
-        // Preserve the shared specialization of a constrained TypeVar. Expanding the TypeVar
-        // before comparing it with `lhs_ty` would lose the correlation between this occurrence
-        // and other occurrences in the same function.
-        if let Type::TypeVar(typevar) = rhs_ty
-            && let Some(TypeVarBoundOrConstraints::Constraints(constraints)) =
-                typevar.typevar(self.db).bound_or_constraints(self.db)
-            && constraints.elements(self.db).iter().all(|constraint| {
-                evaluate_type_equality(self.db, lhs_ty, *constraint, true, soundness_policy)
-                    .is_some_and(|narrowed| narrowed.is_equivalent_to(self.db, *constraint))
-            })
-        {
-            return Some(rhs_ty);
-        }
-
-        evaluate_type_equality(self.db, lhs_ty, rhs_ty, true, soundness_policy)
+        evaluate_type_equality(
+            self.db,
+            lhs_ty,
+            rhs_ty,
+            true,
+            self.comparison_soundness_policy(),
+        )
     }

     fn evaluate_expr_not_in(&self, lhs_ty: Type<'db>, rhs_ty: Type<'db>) -> Option<Type<'db>> {
```

**Comment by AlexWaygood** [comment]
> [PR_26982_ECOSYSTEM_SUMMARY.md](https://github.com/user-attachments/files/30170166/PR_26982_ECOSYSTEM_SUMMARY.md)

FWIW: the skill instructs codex to use GitHub-flavoured Markdown, so these _should_ render really well if you paste the content straight into a PR comment inside a `<details>` block!
